components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/token/handlers/grant/AbstractAuthorizationGrantHandler.java

```diff
@@ -461,9 +468,29 @@ private OAuth2AccessTokenRespDTO generateNewAccessToken(OAuthTokenReqMessageCont
         return createResponseWithTokenBean(newTokenBean, newTokenBean.getValidityPeriodInMillis(), scope);
     }
 
-    private void executePreIssueAccessTokenActions(OAuthTokenReqMessageContext tokenReqMessageContext)
-            throws IdentityOAuth2Exception {
+    private OAuth2AccessTokenRespDTO getFailureOrErrorResponseDTO(ActionExecutionStatus<?> executionStatus) {
+
+        OAuth2AccessTokenRespDTO accessTokenResponse = new OAuth2AccessTokenRespDTO();
+        accessTokenResponse.setError(true);
+        if (executionStatus.getStatus() == ActionExecutionStatus.Status.FAILED) {
+            Failure failureResponse = (Failure) executionStatus.getResponse();
+            accessTokenResponse.setErrorCode(failureResponse.getReason());
+            accessTokenResponse.setErrorMsg(failureResponse.getDescription());
+        } else if (executionStatus.getStatus() == ActionExecutionStatus.Status.ERROR) {
+            Error errorResponse = (Error) executionStatus.getResponse();
+            accessTokenResponse.setErrorCode(StringUtils.isNotBlank(errorResponse.getError()) ?
+                    errorResponse.getError() : OAuth2ErrorCodes.SERVER_ERROR);
+            accessTokenResponse.setErrorMsg(StringUtils.isNotBlank(errorResponse.getDescription()) ?
+                    errorResponse.getDescription() :
+                    ConfigurationConstants.STATUS_INTERNAL_SERVER_ERROR_MESSAGE_DEFAULT);
+        }
```

Comment:
These null checks are not needed if action execution always sends proper error messages. IMO, that contract should hold and we should have a unit test to verify that contract is held